docs/architecture/adr-067-mempool-refactor.md

```diff
@@ -0,0 +1,212 @@
+e.g. [cosmos/cosmos-sdk#8224](https://github.com/cosmos/cosmos-sdk/discussions/8224).
+
+Thus, Tendermint Core needs the ability for an application and its users to
+prioritize transactions in a flexible and performant manner. Specifically, we're
+aiming to either improve, maintain or add the following properties in the
+Tendermint mempool:
+
+- Allow application-determined transaction priority.
+- Allow efficient concurrent reads and writes.
+- Allow block proposers to reap transactions efficiently by priority.
+- Maintain a fixed mempool capacity by transaction size and evict lower priority
+  transactions to make room for higher priority transactions.
+- Allow transactions to be gossiped by priority efficiently.
+- Allow operators to specify a maximum TTL for transactions in the mempool before
+  they're automatically evicted if not selected for a block proposal in time.
+- Ensure the design allows for future extensions, such as replace-by-priority and
+  allowing multiple pending transactions per sender, to be incorporated easily.
+
+Note, not all of these properties will be addressed by the proposed changes in
+this ADR. However, this proposal will ensure that any unaddressed properties
+can be addressed in an easy and extensible manner in the future.
+
+## Alternative Approaches
+
+When considering which approach to take for a priority-based flexible and
+performant mempool, there are two core candidates. The first candidate in less
+invasive in the required  set of protocol and implementation changes, which
+simply extends the existing `CheckTx` ABCI method. The second candidate essentially
+involves the introduction of new ABCI method(s) and would require a higher degree
+of complexity in protocol and implementation changes, some of which may either
+overlap or conflict with the upcoming introduction of [ABCI++](https://github.com/tendermint/spec/blob/master/rfc/004-abci%2B%2B.md).
+
+For more information on the various approaches and proposals, please see the
+[mempool discussion](https://github.com/tendermint/tendermint/discussions/6295).
+
+## Decision
+
+To incorporate a priority-based flexible and performant mempool in Tendermint Core,
+we will introduce new fields, `priority` and `sender`, into the `ResponseCheckTx`
+type and augment the existing mempool data structure to facilitate prioritization
+of uncommitted transactions in addition to extended functionality such as
+replace-by-priority and allowing multiple transactions to exist from the same
+sender with varying priorities.
+
+## Detailed Design
+
+### CheckTx
+
+We introduce the following new fields into the `ResponseCheckTx` type:
+
+```diff
+message ResponseCheckTx {
+  uint32         code       = 1;
+  bytes          data       = 2;
+  string         log        = 3;  // nondeterministic
+  string         info       = 4;  // nondeterministic
+  int64          gas_wanted = 5 [json_name = "gas_wanted"];
+  int64          gas_used   = 6 [json_name = "gas_used"];
```

Comment:
I do not know off the top of my head to be honest.

Reply:
It seems only gas wanted gets used which is to ensure that the gas in a block doesn't exceed `ConsensusParams.Block.MaxGas`. I don't think GasUsed gets used at all.

Reply:
To use gas_wanted for calculating the Block.MaxGas doesn't make sense to me. The reason is if someone assigned transactions with a huge  gas_wanted. It will let the block can only accommodate few transactions. But actually, the proposer can proceed with more transactions in a new block.

Reply:
In principle the fee should be proportional to the gas wanted. You're right it's a bit weird, but because we don't execute txs fully before committing them, we don't know the full gas used until after the block is committed. I believe ProcessProposal of the new ABCI++ will address this kind of thing.